### PR TITLE
Migrate `@woocommerce/date` to TS

### LINKED ISSUE #33100: [Enhancement]: Migrate `@woocommerce/date` to TS
### Describe the solution you'd like

convert @woocommerce/date to TypeScript and output type definitions

### Describe alternatives you've considered

_No response_

### Additional context

Part of https://github.com/woocommerce/woocommerce/issues/32179

### PR DESCRIPTION
**Pull request by chihsuan**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #33100.

This PR migrates @woocommerce/number to TS and 

- fixes the wrong JSDoc types
- adds missing deps `moment-timezone` to fix .tz() type undefined error

### How to test the changes in this Pull Request:

1. Run `pnpm nx build @woocommerce/date`
2. Observe that packages/js/date/build-types` is generated
3. Should run `pnpm nx test @woocommerce/date` without errors

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/date/src/index.ts [hunk 1]
```diff
@@ -567,12 +697,12 @@ export function getIntervalForQuery(
...
  *
- * @param {Object} query           Current query
+ * @param {Query}  query           Current query
  * @param {string} query.chartType
  * @return {string} Current chart type.
  */
-export function getChartTypeForQuery( { chartType } ) {
-	if ( [ 'line', 'bar' ].includes( chartType ) ) {
+export function getChartTypeForQuery( { chartType }: Query ) {
+	if ( chartType !== undefined && [ 'line', 'bar' ].includes( chartType ) ) {
```
Comment (chihsuan) [inline packages/js/date/src/index.ts:714]: `includes` only accept string type.

```ts
Array<string>.includes(searchElement: string, fromIndex?: number | undefined): boolean
```

**Comment by botwoo** [comment]
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33108/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Update tests fail messages` 6c4674fe9ab8445a07a1d609f6d8308fcc0bd86a
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

File: packages/js/date/src/index.ts [hunk 2]
```diff
@@ -745,6 +851,30 @@ export function getDateFormatsForIntervalPhp( interval, ticks = 0 ) {
...
+/**
+ * Returns date formats for the current interval.
+ *
+ * @param {string} interval      Interval to get date formats for.
+ * @param {number} [ticks]       Number of ticks the axis will have.
+ * @param {Object} [option]      Options
+ * @param {string} [option.type] Date format type, d3 or php, defaults to d3.
+ * @return {string} Current interval.
+ */
+export function getDateFormatsForInterval(
```
Comment (chihsuan) [inline packages/js/date/src/index.ts:872]: Moved this function here to fix `no-use-before-declare` warning.

File: packages/js/date/src/test/index.ts [hunk 3]
```diff
@@ -71,19 +79,25 @@ describe( 'toMoment', () => {
 
 	it( 'should handle isoFormat dates', () => {
 		const myMoment = toMoment( 'YYYY', '2018-04-15' );
+		if ( myMoment === null ) fail( 'it should not reach here' );
```
Comment (ilyasfoo) [inline packages/js/date/src/test/index.ts]: Nitpick but I'd prefer the fail message to be more precise, i.e: "myMoment should not be null" or something like that

**Review by ilyasfoo — DISMISSED** [review]
This looks like quite an effort, thanks for wrangling this out, @chihsuan!

The change also was also significant, i.e: with the added null/undefined checks and error throws, so I've done a bit of testing in the analytics and dashboard to make sure none of the consumers were violating those checks. I found none, so I think we're good!

Just a nitpicky comment on test error message, but pre-approving!

File: packages/js/date/src/test/index.ts (hunk 3 above)
Comment (chihsuan) [reply packages/js/date/src/test/index.ts]: @ilyasfoo Thanks for the review. 🙌  Changed them in 6c4674fe9ab8445a07a1d609f6d8308fcc0bd86a

**Review by ilyasfoo — APPROVED** [review]
Thanks! LGTM 🚢

**Pull request merged**